### Importing the Dataset and Libraries

<ins>Data set can be found:</ins>
* https://archive.ics.uci.edu/ml/datasets/Diabetes+130-US+hospitals+for+years+1999-2008
* https://github.com/zeglam/Diabetes-Patient-Readmission-Classification/blob/master/Diabetes%20Patient%20Readmitted%20Prediction.ipynb



In [1]:
import Data_mappings as DM
import pandas as pd
import numpy as np
import seaborn as sns 
import matplotlib.pyplot as plt
df = pd.read_csv("C:/Users/mhmd/Desktop/Uni/Python/dataset_diabetes/diabetic_dat.csv")
df

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)      ?   
1             149190     55629189        Caucasian  Female  [10-20)      ?   
2              64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3             500364     82442376        Caucasian    Male  [30-40)      ?   
4              16680     42519267        Caucasian    Male  [40-50)      ?   
...              ...          ...              ...     ...      ...    ...   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)      ?   
101762     443847782     74694222  AfricanAmerican  Female  [80-90)      ?   
101763     443854148     41088789        Caucasian    Male  [70-80)      ?   
101764     443857166     31693671        Caucasian  Female  [80-90)      ?   
101765     443867222    175429310        Caucasian    Male  [70-80)      ?   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
2                       1                         1                    7   
3                       1                         1                    7   
4                       1                         1                    7   
...                   ...                       ...                  ...   
101761                  1                         3                    7   
101762                  1                         4                    5   
101763                  1                         1                    7   
101764                  2                         3                    7   
101765                  1                         1                    7   

        time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                      1  ...          No      No                   No   
1                      3  ...          No      Up                   No   
2                      2  ...          No      No                   No   
3                      2  ...          No      Up                   No   
4                      1  ...          No  Steady                   No   
...                  ...  ...         ...     ...                  ...   
101761                 3  ...          No    Down                   No   
101762                 5  ...          No  Steady                   No   
101763                 1  ...          No    Down                   No   
101764                10  ...          No      Up                   No   
101765                 6  ...          No      No                   No   

        glipizide-metformin  glimepiride-pioglitazone  \
0                        No                        No   
1                        No                        No   
2                        No                        No   
3                        No                        No   
4                        No                        No   
...                     ...                       ...   
101761                   No                        No   
101762                   No                        No   
101763                   No                        No   
101764                   No                        No   
101765                   No                        No   

        metformin-rosiglitazone  metformin-pioglitazone  change diabetesMed  \
0                            No                      No      No          No   
1                            No                      No      Ch         Yes   
2                            No                      No      No         Yes   
3                            No                      No      Ch         Yes   
4                            No                      No      Ch         Yes   
...                         ...                     ...     ...         ...   
101761                       No

### Looking deeper into the dataset
__Dataset header__

In [2]:
df.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                 1  ...          No      No                   No   
1                 3  ...          No      Up                   No   
2                 2  ...          No      No                   No   
3                 2  ...          No      Up                   No   
4                 1  ...          No  Steady                   No   

   glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
0                   No                        No                       No   
1                   No                        No                       No   
2                   No                        No                       No   
3                   No                        No                       No   
4                   No                        No                       No   

   metformin-pioglitazone  change diabetesMed readmitted  
0                      No      No          No         NO  
1                      No      Ch         Yes        >30  
2                      No      No         Yes         NO  
3                      No      Ch         Yes         NO  
4                      No      Ch         Yes         NO  

[5 rows x 50 columns]

### Cheking the info of the dataset

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

By checking the info we can conclude few things:
* The data contains 101766 entries (rows)
* The **readmitted** column is the column will be using to know the classification
* The type of our features are <ins>integer</ins> or <ins>Object</ins>  


In [4]:
#names of features
print(list(df))

['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight', 'admission_type_id', 'discharge_disposition_id', 'admission_source_id', 'time_in_hospital', 'payer_code', 'medical_specialty', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1', 'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted']


## Data Cleaning
#### We will remove the duplicate Data points
we will use **patient_nbr** as a reference since it has a unique numbers.

In [6]:
df['patient_nbr'].value_counts()

88785891     40
43140906     28
1660293      23
88227540     23
23199021     23
             ..
11005362      1
98252496      1
1019673       1
13396320      1
175429310     1
Name: patient_nbr, Length: 71518, dtype: int64

Out of **101,766** we find that only **71518** are unique <br>
let's try to keep the unique records

In [9]:
df = df.drop_duplicates(subset=['patient_nbr'])
df

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)      ?   
1             149190     55629189        Caucasian  Female  [10-20)      ?   
2              64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3             500364     82442376        Caucasian    Male  [30-40)      ?   
4              16680     42519267        Caucasian    Male  [40-50)      ?   
...              ...          ...              ...     ...      ...    ...   
101754     443842016    183087545        Caucasian  Female  [70-80)      ?   
101755     443842022    188574944            Other  Female  [40-50)      ?   
101756     443842070    140199494            Other  Female  [60-70)      ?   
101758     443842340    120975314        Caucasian  Female  [80-90)      ?   
101765     443867222    175429310        Caucasian    Male  [70-80)      ?   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
2                       1                         1                    7   
3                       1                         1                    7   
4                       1                         1                    7   
...                   ...                       ...                  ...   
101754                  1                         1                    7   
101755                  1                         1                    7   
101756                  1                         1                    7   
101758                  1                         1                    7   
101765                  1                         1                    7   

        time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                      1  ...          No      No                   No   
1                      3  ...          No      Up                   No   
2                      2  ...          No      No                   No   
3                      2  ...          No      Up                   No   
4                      1  ...          No  Steady                   No   
...                  ...  ...         ...     ...                  ...   
101754                 9  ...          No  Steady                   No   
101755                14  ...          No      Up                   No   
101756                 2  ...          No  Steady                   No   
101758                 5  ...          No      Up                   No   
101765                 6  ...          No      No                   No   

        glipizide-metformin  glimepiride-pioglitazone  \
0                        No                        No   
1                        No                        No   
2                        No                        No   
3                        No                        No   
4                        No                        No   
...                     ...                       ...   
101754                   No                        No   
101755                   No                        No   
101756                   No                        No   
101758                   No                        No   
101765                   No                        No   

        metformin-rosiglitazone  metformin-pioglitazone  change diabetesMed  \
0                            No                      No      No          No   
1                            No                      No      Ch         Yes   
2                            No                      No      No         Yes   
3                            No                      No      Ch         Yes   
4                            No                      No      Ch         Yes   
...                         ...                     ...     ...         ...   
101754                       No

Let's run some tests on our different features and see if some columns can be removed 

In [65]:
numberPresentRows =  df.loc[(df['weight'] != '?')].weight.count()
numberMissingRows = len(df.index)
print("percentage of missing rows of {} {} %".format(drug,100-((numberPresentRows/numberMissingRows)*100)//1))

percentage of missing rows of metformin 97.0 %


we can conclude that the weight feature is almost negligible<br>
likewise we will do the same things on different columns <br>
Let's check what are the column drugs that we should remove: 

In [69]:
numberPresentRows =  df.loc[(df['weight'] != '?'),'weight']
numberMissingRows = len(df.index)
print(len(numberPresentRows))
#print("percentage of missing rows of {} {} %".format(numberPresentRows,100-((numberPresentRows/numberMissingRows)*100)//1))

2853


In [74]:
drugslist = ['metformin', 'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'insulin', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone']
for i in range(len(drugslist)):
    numberPresentRows =  len(df.loc[(df[drugslist[i]] != 'No'),drugslist[i]])
    numberMissingRows = len(df.index) 
    print( "percentage of patients not using the drug of {} {} %".format(drugslist[i],100-((numberPresentRows/numberMissingRows)*100)//1))

percentage of patients not using the drug of metformin 80.0 %
percentage of patients not using the drug of repaglinide 99.0 %
percentage of patients not using the drug of nateglinide 100.0 %
percentage of patients not using the drug of chlorpropamide 100.0 %
percentage of patients not using the drug of glimepiride 95.0 %
percentage of patients not using the drug of acetohexamide 100.0 %
percentage of patients not using the drug of glipizide 88.0 %
percentage of patients not using the drug of glyburide 90.0 %
percentage of patients not using the drug of tolbutamide 100.0 %
percentage of patients not using the drug of pioglitazone 93.0 %
percentage of patients not using the drug of rosiglitazone 94.0 %
percentage of patients not using the drug of acarbose 100.0 %
percentage of patients not using the drug of miglitol 100.0 %
percentage of patients not using the drug of troglitazone 100.0 %
percentage of patients not using the drug of tolazamide 100.0 %
percentage of patients not using the

We can see that a lot of the drugs are included in our dataset yet not really used by the patients <br>
meaning that it doesn't really affect the our target **readdmited** column<br>
we will remove every drug that has less that < 5% impact.
we will also remove every column that doesn't affect our work directly such as <ins> incounter_id </ins> ,<ins>patient_nbr</ins>,<ins>payer_code</ins>

In [76]:
columns_drop_list = ['encounter_id', 'patient_nbr', 'weight', 'payer_code','repaglinide', 'nateglinide', 'chlorpropamide','acetohexamide','tolbutamide', 'acarbose', 'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton', 'glyburide-metformin', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone']
df.drop(columns_drop_list,axis = 1, inplace=True)

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 71518 entries, 0 to 101765
Data columns (total 30 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   race                      71518 non-null  object
 1   gender                    71518 non-null  object
 2   age                       71518 non-null  object
 3   admission_type_id         71518 non-null  int64 
 4   discharge_disposition_id  71518 non-null  int64 
 5   admission_source_id       71518 non-null  int64 
 6   time_in_hospital          71518 non-null  int64 
 7   medical_specialty         71518 non-null  object
 8   num_lab_procedures        71518 non-null  int64 
 9   num_procedures            71518 non-null  int64 
 10  num_medications           71518 non-null  int64 
 11  number_outpatient         71518 non-null  int64 
 12  number_emergency          71518 non-null  int64 
 13  number_inpatient          71518 non-null  int64 
 14  diag_1               

In [79]:
df['diag_1']

0         250.83
1            276
2            648
3              8
4            197
           ...  
101754       574
101755       592
101756       996
101758       292
101765       530
Name: diag_1, Length: 71518, dtype: object